### PR TITLE
[Backport] gpstop does not work if previous gpstop was interrupted.

### PR DESCRIPTION
**Pull request by sruthicp**
This is a backport of https://github.com/greenplum-db/gpdb/pull/15727

Issue:
When a gpstop process with 'smart' option is interrupted, the subsequent gpstop command becomes stuck with an error message stating "the database system is shutting down". This error message also appears when attempting to connect to the database using psql and other utilities.
The problem is that the error message does not provide clear instructions on how to stop the database, leading to confusion and an inability to perform a graceful shutdown. Gpstop will not work at this point and the only way to stop the database is to manually kill the master processes.

RCA
The issue occurs due to the interruption of the gpstop process, which leaves the database system in an incomplete shutdown state with unfinished master processes. Also, this will not allow any further operations on the database until the shutdown is complete.

Solution:
This issue will be addressed by modifying the gpstop code to check if the database is shutting down. If yes, it will show a warning stating "The database is currently in the process of shutting down". After this warning, the code will prompt the user to choose between a "Fast" or "Immediate" shutdown option. Depending on the user's selection, the code will proceed as follows:

Terminate all master processes stuck in shutdown state from previously interrupted gpstop.
Start the cluster in master-only mode so that we can get the cluster information required in subsequent steps
Continue the normal gpstop workflow.
Other changes include

adding a 'pidlockpath' in mainOptions of gpstop. This creates a gpstop.lock file under MASTER_DATA_DIRECTORY to prevent the user from running multiple instances of gpstop.
Because of the above change there was one error in pg_upgrade step ("Environment Variable MASTER_DATA_DIRECTORY not set!"). Added fix for that as well.

Tests done:
Added new unit test cases and behave test cases for testing the scenario

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/bin/gpstop [hunk 1]
```diff
@@ -1028,6 +1085,17 @@ class GpStop:
...
+        the customer from running multiple instances of gpstop.
+        """
+        return {'pidlockpath': 'gpstop.lock', 'parentpidvar': 'GPSTOPPID'}
 
 if __name__ == '__main__':
-    simple_main(GpStop.createParser, GpStop.createProgram)
+    simple_main(GpStop.createParser, GpStop.createProgram, GpStop.mainOptions())
+
+
+
```
Comment (jnihal) [inline gpMgmt/bin/gpstop:1101]: nit: Remove the extra lines

**Review by ravoorsh — APPROVED** [review]
LGTM+1

**Review by nimish350 — APPROVED** [review]
LGTM +1

**Review by jnihal — APPROVED** [review]

**Comment by ashwinstar** [comment]
Seems neither main branch PR nor this one updated the help section of the utility. It currently says...
```
-M smart
 
 Smart shut down. If there are active connections, this command 
 fails with a warning. This is the default shutdown mode.
```
Current documentation/help is anyways wrong without this PR as well given it waits for user connections and doesn't really just fail with warning. So, please help to make it reflect the new reality and behavior. Thanks.

**Comment by nimish350** [comment]
> 

@ashwinstar The docs for GP7 and latest GP6 reflect the change. Let me know if we need: 

```
-M smart
Smart shut down. This is the default shutdown mode. gpstop waits for active user connections to disconnect and then proceeds with the shutdown. If any user connections remain open after the timeout period (or if you interrupt by pressing CTRL-C) gpstop lists the open user connections and prompts whether to continue waiting for connections to finish, or to perform a fast or immediate shutdown.
```

> Seems neither main branch PR nor this one updated the help section of the utility. It currently says...
> 
> ```
> -M smart
>  
>  Smart shut down. If there are active connections, this command 
>  fails with a warning. This is the default shutdown mode.
> ```
> 
> Current documentation/help is anyways wrong without this PR as well given it waits for user connections and doesn't really just fail with warning. So, please help to make it reflect the new reality and behavior. Thanks.

Discussed this with @sruthicp. We need to change the gpstop --help to reflect the change that has been updated in the Docs: 
```
-M smart
Smart shut down. This is the default shutdown mode. gpstop waits for active user connections to disconnect and then proceeds with the shutdown. If any user connections remain open after the timeout period (or if you interrupt by pressing CTRL-C) gpstop lists the open user connections and prompts whether to continue waiting for connections to finish, or to perform a fast or immediate shutdown.
```

**Comment by sruthicp** [comment]
> > 
> 
> @ashwinstar The docs for GP7 and latest GP6 reflect the change. Let me know if we need:
> 
> ```
> -M smart
> Smart shut down. This is the default shutdown mode. gpstop waits for active user connections to disconnect and then proceeds with the shutdown. If any user connections remain open after the timeout period (or if you interrupt by pressing CTRL-C) gpstop lists the open user connections and prompts whether to continue waiting for connections to finish, or to perform a fast or immediate shutdown.
> ```
> 
> > Seems neither main branch PR nor this one updated the help section of the utility. It currently says...
> > ```
> > -M smart
> >  
> >  Smart shut down. If there are active connections, this command 
> >  fails with a warning. This is the default shutdown mode.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Current documentation/help is anyways wrong without this PR as well given it waits for user connections and doesn't really just fail with warning. So, please help to make it reflect the new reality and behavior. Thanks.
> 
> Discussed this with @sruthicp. We need to change the gpstop --help to reflect the change that has been updated in the Docs:
> 
> ```
> -M smart
> Smart shut down. This is the default shutdown mode. gpstop waits for active user connections to disconnect and then proceeds with the shutdown. If any user connections remain open after the timeout period (or if you interrupt by pressing CTRL-C) gpstop lists the open user connections and prompts whether to continue waiting for connections to finish, or to perform a fast or immediate shutdown.
> ```

@ashwinstar @nimish350 Updated the doc for the gpstop --help.

**Comment by ashwinstar** [comment]
> Updated the doc for the gpstop --help.

Thanks reads good to me.

**Pull request merged**